In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# What is Data Cleaning?

 Data Cleaning is a process of fixing or removing incorrect, corrupted, incorrectly formated, duplicate or incomplete data with in the dataset. 

In [2]:
# Importing Dataset

dataset = pd.read_csv("Car_Price.csv")
dataset.head(5)

Unnamed: 0                                car_name car_prices_in_rupee  \
0           0  Jeep Compass 2.0 Longitude Option BSIV          10.03 Lakh   
1           1            Renault Duster RXZ Turbo CVT          12.83 Lakh   
2           2                      Toyota Camry 2.5 G          16.40 Lakh   
3           3                       Honda Jazz VX CVT           7.77 Lakh   
4           4        Volkswagen Polo 1.2 MPI Highline           5.15 Lakh   

   kms_driven fuel_type transmission  ownership  manufacture   engine    Seats  
0  86,226 kms    Diesel       Manual  1st Owner         2017  1956 cc  5 Seats  
1  13,248 kms    Petrol    Automatic  1st Owner         2021  1330 cc  5 Seats  
2  60,343 kms    Petrol    Automatic  1st Owner         2016  2494 cc  5 Seats  
3  26,696 kms    Petrol    Automatic  1st Owner         2018  1199 cc  5 Seats  
4  69,414 kms    Petrol       Manual  1st Owner         2016  1199 cc  5 Seats

In [3]:
# Checking Datatype

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5512 entries, 0 to 5511
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           5512 non-null   int64 
 1   car_name             5512 non-null   object
 2   car_prices_in_rupee  5512 non-null   object
 3   kms_driven           5512 non-null   object
 4   fuel_type            5512 non-null   object
 5   transmission         5512 non-null   object
 6   ownership            5512 non-null   object
 7   manufacture          5512 non-null   int64 
 8   engine               5512 non-null   object
 9   Seats                5512 non-null   object
dtypes: int64(2), object(8)
memory usage: 430.8+ KB


In [4]:
# Deleting Unwanted Column.

dataset = dataset.drop(["Unnamed: 0"],axis=1)          # the code is correct.

In [5]:
dataset.head(1)

car_name car_prices_in_rupee  kms_driven  \
0  Jeep Compass 2.0 Longitude Option BSIV          10.03 Lakh  86,226 kms   

  fuel_type transmission  ownership  manufacture   engine    Seats  
0    Diesel       Manual  1st Owner         2017  1956 cc  5 Seats

In [6]:
dataset.shape

(5512, 9)

In [7]:
dataset.describe().T

count         mean       std     min     25%     50%     75%  \
manufacture  5512.0  2015.455552  3.927974  1995.0  2013.0  2016.0  2018.0   

                max  
manufacture  2022.0

# Handling Missing Values

In [8]:
dataset.isnull().sum()

car_name               0
car_prices_in_rupee    0
kms_driven             0
fuel_type              0
transmission           0
ownership              0
manufacture            0
engine                 0
Seats                  0
dtype: int64

In [9]:
# There are No missing values avalible in this dataset.

In [10]:
# If there was a missing value avalible then the missing value would be removed or filled upward/backward.

# dataset.dropna(inplace = True) ----> To Remove missing value.
# dataset.fillna(method = "ffill")  ----> To fill upward.

# Standardizing Data Formats

In [11]:
# Here we will perform Data Cleaning in all the Columns one by one.

# car_name (column 1)

In [12]:
dataset["car_name"]

0       Jeep Compass 2.0 Longitude Option BSIV
1                 Renault Duster RXZ Turbo CVT
2                           Toyota Camry 2.5 G
3                            Honda Jazz VX CVT
4             Volkswagen Polo 1.2 MPI Highline
                         ...                  
5507                   BMW X1 sDrive 20d xLine
5508                     BMW M Series M4 Coupe
5509                Jaguar XF 2.2 Litre Luxury
5510                        BMW 7 Series 730Ld
5511                 BMW 5 Series 520d M Sport
Name: car_name, Length: 5512, dtype: object

In [13]:
dataset["car_name"][0]

'Jeep Compass 2.0 Longitude Option BSIV'

* Problem Statement :
     * Company name and Car name lies in Same column. so, we need to supperate them in two different columns.

In [14]:
# extracting company name from column.

x = dataset["car_name"][0]
x[:x.index(" ")]            # here the space "" (" ") "" represents the space that comes after company name "" Jeep  "".

# Now the company name is extracted.

'Jeep'

In [15]:
# Now aplly this step in whole dataset.

def company_name(x):
    return x[:x.index(" ")]
dataset["car_name"].apply(company_name)

0             Jeep
1          Renault
2           Toyota
3            Honda
4       Volkswagen
           ...    
5507           BMW
5508           BMW
5509        Jaguar
5510           BMW
5511           BMW
Name: car_name, Length: 5512, dtype: object

In [16]:
# adding "company_name" column to the dataset.  OR   saving the changes to dataset.

dataset["company_name"] = dataset["car_name"].apply(company_name)

In [17]:
dataset.head(1)

car_name car_prices_in_rupee  kms_driven  \
0  Jeep Compass 2.0 Longitude Option BSIV          10.03 Lakh  86,226 kms   

  fuel_type transmission  ownership  manufacture   engine    Seats  \
0    Diesel       Manual  1st Owner         2017  1956 cc  5 Seats   

  company_name  
0         Jeep

In [18]:
# Extracting car_name.

def car_name(x):
    return x[x.index(" ")+1:]
dataset["car_name"].apply(car_name)

0       Compass 2.0 Longitude Option BSIV
1                    Duster RXZ Turbo CVT
2                             Camry 2.5 G
3                             Jazz VX CVT
4                   Polo 1.2 MPI Highline
                      ...                
5507                  X1 sDrive 20d xLine
5508                    M Series M4 Coupe
5509                  XF 2.2 Litre Luxury
5510                       7 Series 730Ld
5511                5 Series 520d M Sport
Name: car_name, Length: 5512, dtype: object

In [19]:
# adding "car_name" column to the dataset. OR  saving the changes to dataset.

dataset["car_name"] = dataset["car_name"].apply(car_name)

In [20]:
dataset.head(2)

car_name car_prices_in_rupee  kms_driven  \
0  Compass 2.0 Longitude Option BSIV          10.03 Lakh  86,226 kms   
1               Duster RXZ Turbo CVT          12.83 Lakh  13,248 kms   

  fuel_type transmission  ownership  manufacture   engine    Seats  \
0    Diesel       Manual  1st Owner         2017  1956 cc  5 Seats   
1    Petrol    Automatic  1st Owner         2021  1330 cc  5 Seats   

  company_name  
0         Jeep  
1      Renault

# car_prices_in_rupee (column 2)

In [21]:
dataset["car_prices_in_rupee"]

0       10.03 Lakh
1       12.83 Lakh
2       16.40 Lakh
3        7.77 Lakh
4        5.15 Lakh
           ...    
5507    28.90 Lakh
5508    64.90 Lakh
5509    13.75 Lakh
5510    29.90 Lakh
5511    31.90 Lakh
Name: car_prices_in_rupee, Length: 5512, dtype: object

* Problem Statement :
   * The Data is not in number format. it is in mixed format like decimal, coma, string etc. Data should be in same format.
   * Mistakes :- 35,000  ,  28.90 Lakh  , 1.02 Crore. All the formats are different

In [22]:
# Replacing or Removing coma (35,000)

dataset["car_prices_in_rupee"].str.replace(",","")

0       10.03 Lakh
1       12.83 Lakh
2       16.40 Lakh
3        7.77 Lakh
4        5.15 Lakh
           ...    
5507    28.90 Lakh
5508    64.90 Lakh
5509    13.75 Lakh
5510    29.90 Lakh
5511    31.90 Lakh
Name: car_prices_in_rupee, Length: 5512, dtype: object

In [23]:
# saving the changes to dataset.

dataset["car_prices_in_rupee"] = dataset["car_prices_in_rupee"].str.replace(",","")

In [24]:
# Convert Lakh and Crore to Thousands.

def rupee_change(x):
    p = x.split(" ")
    try:
        if p[1] == "Lakh":
            return str(round(float(p[0])*100000,1))
        elif p[1] == "Crore":
            return str(round(float(p[0])*10000000,1))
    except:
        return x

In [25]:
dataset["car_prices_in_rupee"].apply(rupee_change)

0       1003000.0
1       1283000.0
2       1640000.0
3        777000.0
4        515000.0
          ...    
5507    2890000.0
5508    6490000.0
5509    1375000.0
5510    2990000.0
5511    3190000.0
Name: car_prices_in_rupee, Length: 5512, dtype: object

In [26]:
# saving the changes to dataset.

dataset["car_prices_in_rupee"] = dataset["car_prices_in_rupee"].apply(rupee_change)

In [27]:
# Changing Datatype to float.

dataset["car_prices_in_rupee"].astype("float")

0       1003000.0
1       1283000.0
2       1640000.0
3        777000.0
4        515000.0
          ...    
5507    2890000.0
5508    6490000.0
5509    1375000.0
5510    2990000.0
5511    3190000.0
Name: car_prices_in_rupee, Length: 5512, dtype: float64

In [28]:
# saving the changes to dataset.

dataset["car_prices_in_rupee"] = dataset["car_prices_in_rupee"].astype("float")

In [29]:
dataset.info()    # Datatype is changed to float.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5512 entries, 0 to 5511
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   car_name             5512 non-null   object 
 1   car_prices_in_rupee  5389 non-null   float64
 2   kms_driven           5512 non-null   object 
 3   fuel_type            5512 non-null   object 
 4   transmission         5512 non-null   object 
 5   ownership            5512 non-null   object 
 6   manufacture          5512 non-null   int64  
 7   engine               5512 non-null   object 
 8   Seats                5512 non-null   object 
 9   company_name         5512 non-null   object 
dtypes: float64(1), int64(1), object(8)
memory usage: 430.8+ KB


# kms_driven (column 3)

In [30]:
dataset["kms_driven"]

0       86,226 kms
1       13,248 kms
2       60,343 kms
3       26,696 kms
4       69,414 kms
           ...    
5507    45,000 kms
5508    29,000 kms
5509    90,000 kms
5510    79,000 kms
5511    42,000 kms
Name: kms_driven, Length: 5512, dtype: object

* Problem Statement : 
  * Datatype is in Object which is incorrect.So, by removing coma we have to convert datatype from object to float.
  * "kms" should be removed.

In [31]:
# removing coma.

dataset["kms_driven"].str.replace(",","")

0       86226 kms
1       13248 kms
2       60343 kms
3       26696 kms
4       69414 kms
          ...    
5507    45000 kms
5508    29000 kms
5509    90000 kms
5510    79000 kms
5511    42000 kms
Name: kms_driven, Length: 5512, dtype: object

In [32]:
# saving the changes to dataset.

dataset["kms_driven"] = dataset["kms_driven"].str.replace(",","")

In [33]:
# Removing "kms".

dataset["kms_driven"].str.replace(" kms","")

0       86226
1       13248
2       60343
3       26696
4       69414
        ...  
5507    45000
5508    29000
5509    90000
5510    79000
5511    42000
Name: kms_driven, Length: 5512, dtype: object

In [34]:
# saving the changes to dataset.

dataset["kms_driven"] = dataset["kms_driven"].str.replace(" kms","")

In [35]:
# Changing Datatype from Object to Int.

dataset["kms_driven"].astype("int64")                # here "int64" reffers to 64bit, our system supports 64bit thats why we kept int as int64.

0       86226
1       13248
2       60343
3       26696
4       69414
        ...  
5507    45000
5508    29000
5509    90000
5510    79000
5511    42000
Name: kms_driven, Length: 5512, dtype: int64

In [36]:
# saving the changes to dataset.

dataset["kms_driven"] = dataset["kms_driven"].astype("int64")

In [37]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5512 entries, 0 to 5511
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   car_name             5512 non-null   object 
 1   car_prices_in_rupee  5389 non-null   float64
 2   kms_driven           5512 non-null   int64  
 3   fuel_type            5512 non-null   object 
 4   transmission         5512 non-null   object 
 5   ownership            5512 non-null   object 
 6   manufacture          5512 non-null   int64  
 7   engine               5512 non-null   object 
 8   Seats                5512 non-null   object 
 9   company_name         5512 non-null   object 
dtypes: float64(1), int64(2), object(7)
memory usage: 430.8+ KB


# fuel_type (column 4)

In [38]:
dataset["fuel_type"]

0       Diesel
1       Petrol
2       Petrol
3       Petrol
4       Petrol
         ...  
5507    Diesel
5508    Petrol
5509    Diesel
5510    Diesel
5511    Diesel
Name: fuel_type, Length: 5512, dtype: object

In [39]:
# The Data is Correct.

# transmission (column 5)

In [40]:
dataset["transmission"]

0          Manual
1       Automatic
2       Automatic
3       Automatic
4          Manual
          ...    
5507    Automatic
5508    Automatic
5509    Automatic
5510    Automatic
5511    Automatic
Name: transmission, Length: 5512, dtype: object

In [41]:
# The data is Correct.

# ownership (column 6)

In [42]:
dataset["ownership"]

0       1st Owner
1       1st Owner
2       1st Owner
3       1st Owner
4       1st Owner
          ...    
5507    1st Owner
5508    2nd Owner
5509    2nd Owner
5510    3rd Owner
5511    2nd Owner
Name: ownership, Length: 5512, dtype: object

* Problem Statement : 
   * 

In [43]:
# Leave the "ownership" for later.

# manufacture (column 7)

In [44]:
dataset["manufacture"]

0       2017
1       2021
2       2016
3       2018
4       2016
        ... 
5507    2018
5508    2015
5509    2013
5510    2015
5511    2017
Name: manufacture, Length: 5512, dtype: int64

In [45]:
# The data is correct.

# engine (column 8)

In [46]:
dataset["engine"]

0       1956 cc
1       1330 cc
2       2494 cc
3       1199 cc
4       1199 cc
         ...   
5507    2995 cc
5508    1968 cc
5509    2755 cc
5510    2967 cc
5511    1991 cc
Name: engine, Length: 5512, dtype: object

* Problem Statement :
   * "cc" should be removed.
   * Convert datatype from Object to Int64.

In [47]:
# removing "cc".

dataset["engine"].str.replace(" cc","")            # the code is correct.

0       1956
1       1330
2       2494
3       1199
4       1199
        ... 
5507    2995
5508    1968
5509    2755
5510    2967
5511    1991
Name: engine, Length: 5512, dtype: object

In [48]:
# saving the changes to dataset.

dataset["engine"] = dataset["engine"].str.replace(" cc","")    # The code is correct.

In [49]:
# Converting datatype from object to int.

dataset["engine"].astype("int64")

0       1956
1       1330
2       2494
3       1199
4       1199
        ... 
5507    2995
5508    1968
5509    2755
5510    2967
5511    1991
Name: engine, Length: 5512, dtype: int64

In [50]:
# saving the changes to dataset.

dataset["engine"] = dataset["engine"].astype("int64")

# Seats (column 9)

In [51]:
dataset["Seats"]

0       5 Seats
1       5 Seats
2       5 Seats
3       5 Seats
4       5 Seats
         ...   
5507    7 Seats
5508    5 Seats
5509    5 Seats
5510    6 Seats
5511    5 Seats
Name: Seats, Length: 5512, dtype: object

* Problem Statement : 
    * Remove " Seats".
    * Convert datatype to int64.

In [52]:
# Removing " Seats". 

dataset["Seats"].str.replace(" Seats","")

0       5
1       5
2       5
3       5
4       5
       ..
5507    7
5508    5
5509    5
5510    6
5511    5
Name: Seats, Length: 5512, dtype: object

In [53]:
# saving the changes to dataset.

dataset["Seats"] = dataset["Seats"].str.replace(" Seats","")

In [54]:
# Converting Datatype from Object to Int64.

dataset["Seats"].astype("int64")

0       5
1       5
2       5
3       5
4       5
       ..
5507    7
5508    5
5509    5
5510    6
5511    5
Name: Seats, Length: 5512, dtype: int64

In [55]:
# saving the changes to dataset.

dataset["Seats"] = dataset["Seats"].astype("int64")

# company_name(column 10)

In [56]:
dataset["company_name"]

0             Jeep
1          Renault
2           Toyota
3            Honda
4       Volkswagen
           ...    
5507           BMW
5508           BMW
5509        Jaguar
5510           BMW
5511           BMW
Name: company_name, Length: 5512, dtype: object

In [57]:
# The data is correct.

*NOTE : The mistake is not avalible in this dataset but need to have knowledge about that.

* Suppose the Date is not in correct format.for example "manufacture_date" is a column and it's format is in 7 january,2022 format. which is not correct. So, to solve this we will write following code:

    * pd.to_datetime(dataset["manufacture_date"])  # The date format will be corrected eg: 2022-01-07.
    * dataset["manufacture_date"] = pd.to_datetime(dataset["manufacture_date"])  # save the changes to dataset.

In [58]:
dataset.head(5)

car_name  car_prices_in_rupee  kms_driven  \
0  Compass 2.0 Longitude Option BSIV            1003000.0       86226   
1               Duster RXZ Turbo CVT            1283000.0       13248   
2                        Camry 2.5 G            1640000.0       60343   
3                        Jazz VX CVT             777000.0       26696   
4              Polo 1.2 MPI Highline             515000.0       69414   

  fuel_type transmission  ownership  manufacture  engine  Seats company_name  
0    Diesel       Manual  1st Owner         2017    1956      5         Jeep  
1    Petrol    Automatic  1st Owner         2021    1330      5      Renault  
2    Petrol    Automatic  1st Owner         2016    2494      5       Toyota  
3    Petrol    Automatic  1st Owner         2018    1199      5        Honda  
4    Petrol       Manual  1st Owner         2016    1199      5   Volkswagen

# Now this Data is ready for Analysis.

# Filling unwanted outliers.

* In this Dataset there is no unwanted outlier avaliable.  
* NOTE: to see outlier the data must contain numeric value.

* To see outlier we can use boxplot. 
    *     eg:  plt.figure(figsize = (4,5))
    *          sns.boxplot(x="column_name",data=dataset)
    *          plt.show()